### PR TITLE
Add timetable

### PR DESCRIPTION
**Pull request by doteq**

### CODE REVIEW

File: src/diary/diary.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import format from 'date-fns/format'
+import { UserObject } from './interfaces/UserObject';
+import { TimetableParser } from './parsers/timetableParser';
+
+const axios = require('axios').default;
+const axiosCookieJarSupport = require('axios-cookiejar-support').default;
+
+let api;
+
+export class Diary {
```
Comment (dominik-korsa) [inline src/diary/diary.ts]: Please rename `Diary` to `Register`.
https://pl.bab.la/slownik/polski-angielski/dziennik-szkolny

File: src/client/client.ts [hunk 2]
```diff
@@ -73,11 +73,30 @@ export class Client extends BaseClient {
   private async initClient(mainResponse: string, symbol: string): Promise<boolean> {
     this.symbol = symbol;
     const $ = cheerio.load(mainResponse);
-    // REMOVE LINE BELOW WHEN CREATING THIS FEATURE
-    // eslint-disable-next-line @typescript-eslint/no-unused-vars
-    const links: string[] = $('.panel.linkownia.pracownik.klient a[href*="uonetplus-uczen"]').map(
+    this.urlList = $('.panel.linkownia.pracownik.klient a[href*="uonetplus-uczen"]').map(
       (index, value) => $(value).attr('href'),
     ).get();
     return true;
```
Comment (dominik-korsa) [inline src/client/client.ts]: What is the point of the boolean return value? When can the function return `false`?
If you don't want to return any value use type `Promise<void>`.

File: src/client/client.ts [hunk 3]
```diff
@@ -73,11 +73,30 @@ export class Client extends BaseClient {
...
   }
+
+  public async getDiaryList(): Promise<any[]> {
```
Comment (dominik-korsa) [inline src/client/client.ts]: Rename to `getRegisterList` or `getRegistersList`.
https://stackoverflow.com/questions/9784720/items-list-or-item-list

File: src/diary/diary.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    this.cookieJar.setCookie(`idBiezacyUczen=${this.userObject.IdUczen}; path=/; domain=uonetplus-uczen.${this.userObject.host}`, `https://uonetplus-uczen.${this.userObject.host}`, () => {});
+    api = axios.create({
+      baseURL: userObject.url,
+      headers: { jar: this.cookieJar },
+    });
+    axiosCookieJarSupport(api);
+  }
+
+  public getTimetable(date: Date): Promise<object> {
+    console.log(this.getWeekDateString(date));
```
Comment (dominik-korsa) [inline src/diary/diary.ts]: Do not use console.log in final code

File: src/diary/interfaces/UserObject.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export interface UserObject {
+  IdDziennik: number;
+  IdUczen: number;
+  url: string;
+  host: string;
+}
```
Comment (dominik-korsa) [inline src/diary/interfaces/UserObject.ts]: Rename file to `user-object.ts`

File: src/diary/interfaces/timetable/TimetableHeaders.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+export interface TimetableHeaders {
+  Text: string;
+  Width: string;
+  Distinction: boolean;
+  Flex: number;
+}
```
Comment (dominik-korsa) [inline src/diary/interfaces/timetable/TimetableHeaders.ts]: Rename file to `timetable-headers.ts`

File: src/diary/interfaces/timetable/TimetableLesson.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  group: string;
+  room: string;
+  roomOld?: string;
+  teacher: string;
+  teacherOld?: string;
+  info: string;
+  studentPlan?: boolean;
+  changes: boolean;
+  canceled: boolean;
+}
```
Comment (dominik-korsa) [inline src/diary/interfaces/timetable/TimetableLesson.ts]: Rename file to `timetable-lesson.ts`

File: src/diary/interfaces/timetable/TimetableResponse.ts [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import { TimetableHeaders } from './TimetableHeaders';
+
+export interface TimetableResponse {
+  Data: string;
+  Headers: TimetableHeaders[];
+  Rows: string[][];
+  Additionals: string[];
+}
```
Comment (dominik-korsa) [inline src/diary/interfaces/timetable/TimetableResponse.ts]: Rename file to `timetable-response.ts`

File: tests/client.e2e.ts [hunk 9]
```diff
@@ -1,9 +1,22 @@
...
+      diaryList = await client.getDiaryList();
+      expect(diaryList[0].host).toEqual('fakelog.cf');
+    });
+  });
+});
+
+
+describe('Diary', () => {
+  it('Get timetable', async () => {
+    const diary = new wulkanowy.Diary(diaryList[0], client.cookieJar);
```
Comment (dominik-korsa) [inline tests/client.e2e.ts]: Don't use a variable (`diaryList` in this case), when its value is defined in another test. Tests should be able to run on their own. Please also sync this branch with master, where the tests are slightly changed.

File: src/client/client.ts (hunk 2 above)
Comment (dominik-korsa) [reply src/client/client.ts:79]: A fix for this is in #16

File: src/client/client.ts [hunk 10]
```diff
@@ -73,11 +73,30 @@ export class Client extends BaseClient {
...
+    responses.forEach((response) => {
+      if (response.data.success === true) {
+        for (let i = 0; i < response.data.data.length; i += 1) {
+          response.data.data[i].url = response.config.url
+            .substring(0, response.config.url.length - 22);
+          response.data.data[i].host = this.host;
+        }
+      }
+    });
+    return [].concat(...responses.map((x) => x.data.data));
```
Comment (pniedzwiedzinski) [inline src/client/client.ts:100]: Wouldn't it be simpler?
```suggestion
    responses
      .filter((response) => response.data.success === true)
      .flatMap((response) => // : Array<UserObject>
        response.data.data.map((responseData) => ({ // : UserObject
          IdDziennik: responseData.IdDziennik,
          IdUczen: responseData.IdUczen,
          url: response.config.url.substring(0, response.config.url.length - 22),
          host: this.host,
        }))
      );
    return responses
```

File: src/diary/diary.ts (hunk 1 above)
Comment (doteq) [reply src/diary/diary.ts:12]: wulkanowy/sdk uses diary, wontfix

File: src/client/client.ts (hunk 3 above)
Comment (doteq) [reply src/client/client.ts:84]: Same as above.

**Review by dominik-korsa — APPROVED** [review]

File: src/diary/parsers/timetableParser.ts [hunk 11]
```diff
@@ -0,0 +1,198 @@
+import * as cheerio from 'cheerio';
+import { TimetableLesson } from '../interfaces/timetable/timetable-lesson';
+import { TimetableResponse } from '../interfaces/timetable/timetable-response';
+
+export class TimetableParser {
+  public static parseTimetable(htmlResponse: TimetableResponse): TimetableLesson[] {
+    const timetable: TimetableLesson[] = [];
+    const weekDays = htmlResponse.Headers;
+    weekDays.shift();
```
Comment (dominik-korsa) [inline src/diary/parsers/timetableParser.ts:9]: Do not modify the original array

**Review by dominik-korsa — CHANGES_REQUESTED** [review]

**Review by dominik-korsa — APPROVED** [review]

File: src/client/client.ts [hunk 12]
```diff
@@ -74,8 +80,27 @@ export class Client extends BaseClient {
...
+    this.urlList = $('.panel.linkownia.pracownik.klient a[href*="uonetplus-uczen"]')
+      .toArray()
+      .map((element) => $(element).attr('href'))
+      .filter(notNil);
+  }
+
+  public async getDiaryList(): Promise<UserObject[]> {
+    if (this.urlList === undefined) throw new NoUrlListError();
+    const results = await Promise.all(this.urlList.map(async (studentUrl) => {
+      const url = new URL('UczenDziennik.mvc/Get', studentUrl).toString();
```
Comment (dominik-korsa) [inline src/client/client.ts:92]: Might fail if trailing slash is missing

File: src/diary/diary.ts [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+   */
+  public getTimetable(date: Date): Promise<object> {
+    return new Promise((resolve) => {
+      this.api.post('http://uonetplus-uczen.fakelog.cf/powiatwulkanowy/123456/PlanZajec.mvc/Get', qs.stringify({ date: Diary.getWeekDateString(date) })).then((response) => {
+        if (response.data.success) {
+          resolve(TimetableParser.parseTimetable(response.data.data));
+        }
+      });
+    });
+  }
```
Comment (dominik-korsa) [inline src/diary/diary.ts:47]: Convert to async/await and handle errors

**Review by dominik-korsa — CHANGES_REQUESTED** [review]

**Review by dominik-korsa — APPROVED** [review]

File: src/diary/diary.ts [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+  /**
+   * Api diary for SDK constructor.
+   * @param userObject Selected diary from diary list.
+   * @param cookieJar Client's cookie jar.
+   */
+  public constructor(userObject: UserObject, cookieJar: CookieJar) {
+    this.cookieJar = cookieJar;
+    this.userObject = userObject;
+    this.cookieJar.setCookie(`idBiezacyDziennik=${this.userObject.IdDziennik}; path=/; domain=uonetplus-uczen.${this.userObject.host}`, `https://uonetplus-uczen.${this.userObject.host}`, () => {});
+    this.cookieJar.setCookie(`idBiezacyUczen=${this.userObject.IdUczen}; path=/; domain=uonetplus-uczen.${this.userObject.host}`, `https://uonetplus-uczen.${this.userObject.host}`, () => {});
```
Comment (dominik-korsa) [inline src/diary/diary.ts:30]: This modifies the original cookie jar which might be an unwanted behavior, as it could break using multiple diaries at once

**Review by dominik-korsa — COMMENTED** [review]
**(resolved)** getTimetable throws `Error: Request failed with status code 429`

**Pull request merged**